apps/massa_proxy/lib/massa_proxy/runtime/grpc/protocol/action/stream_handler.ex

```diff
@@ -2,42 +2,120 @@ defmodule MassaProxy.Runtime.Grpc.Protocol.Action.Stream.Handler do
   @moduledoc """
   This module is responsible for handling stream requests of the Action protocol
   """
-  require Logger
-  alias GRPC.Server
+
   alias Google.Protobuf.Any
+  alias Cloudstate.{Action.ActionCommand, Metadata}
+  alias Cloudstate.Action.ActionProtocol.Stub, as: ActionClient
+
+  import MassaProxy.Util, only: [get_connection: 0, get_type_url: 1]
+
+  def handle_streamed(%{stream: stream, message: message} = payload) do
+    messages = build_stream(message, payload)
+
+    with {:ok, conn} <- get_connection(),
+         client_stream = ActionClient.handle_streamed(conn),
+         :ok <- client_stream |> run_stream(messages) |> Stream.run(),
+         {:ok, consumer_stream} <- GRPC.Stub.recv(client_stream) do
+      consumer_stream
+      |> Stream.each(fn {:ok, r} -> GRPC.Server.send_reply(stream, unpack_response(r)) end)
+      |> Stream.run()
+    else
+      {:error, _reason} = err -> err
+    end
+  end
+
+  def handle_stream_in(%{message: message} = payload) do
+    messages = build_stream(message, payload)
+
+    with {:ok, conn} <- get_connection(),
+         client_stream = ActionClient.handle_streamed_in(conn),
+         task_result <- run_stream(client_stream, messages),
+         :ok <- accumlate_stream_result(task_result),
+         {:ok, response} <- GRPC.Stub.recv(client_stream) do
+      unpack_response(response)
+    else
+      {:error, _reason} = err -> err
+    end
+  end
+
+  def handle_stream_out(%{stream: stream} = payload) do
+    message =
+      ActionCommand.new(
+        service_name: payload.service_name,
+        name: payload.original_method,
+        payload:
+          Any.new(
+            type_url: get_type_url(payload.input_type),
+            value: payload.input_type.encode(payload.message)
+          ),
+        metadata: Metadata.new()
+      )
+
+    with {:ok, conn} <- get_connection(),
+         {:ok, client_stream} <- ActionClient.handle_streamed_out(conn, message, []) do
+      Stream.each(client_stream, fn {:ok, response} ->
+        GRPC.Server.send_reply(stream, unpack_response(response))
+      end)
+      |> Stream.run()
+    end
+  end
+
+  defp run_stream(client_stream, messages) do
+    Stream.map(messages, &send_stream_msg(client_stream, &1))
+  end
 
-  def handle_streamed(
-        %{
-          stream: stream,
-          input_type: input_type,
-          output_type: output_type
-        } = payload
-      ) do
-    Stream.each(stream, fn msg ->
-      Logger.info("Decode request from #{inspect(msg)}")
-      handle_streamed_message(stream, msg)
+  defp unpack_response(%{response: {_, response}}), do: response
+
+  defp accumlate_stream_result(results) do
```

Comment:
I don't know if accumulating is what we want here, the ideal would be to have a single response from the user function, we should send the incoming stream to the process function and wait for a single response at the end of the processing of this stream carried out by the user function.
What is your reasoning about this?

Reply:
Yes, agree. However since we are sending chunks to the user function, there's a chance for failure mid-stream. (User function died, network problems etc). So there needs a way to enumerate the stream and to determine if the stream as a whole succeed. Example:

```
grpcurl -d @ --plaintext localhost:9000 cloudstate.tck.model.action.ActionTckModel.ProcessStreamedIn << EOF
{}
{"groups": {"steps": [{"reply":{"message":"AAAAA"}}]}}
{"groups": {"steps": [{"reply":{"message":"BBBBB"}}]}} # fails
{"groups": {"steps": [{"reply":{"message":"CCCCC"}}]}} # Do you still send this one?
EOF
```
So the problem here is what guarantees is provided:

* should the request fail fail in-case of first encountered error? 
* should message 2 be retried before trying to send 3?
* skip message 2 and continue?

The intent here was to fail the *whole* request if any of the streamed messages fails for whatever reason. So it needs some kind of processing state to determine the current request and to abort the enumeration if a request failed. `accumulate_stream_result` is probably a poor name as it only keeps the actual state of the stream processing.

One could just build the `Stream` pipeline and do `Stream.run` but it would be hard to determine if it was a successfull/failed request.

Reply:
I think it might work the way you did. We have to test this a lot with the TCK